### ISSUE TITLE
Improve "Unable to retrieve the start stamps.  Did you just reshard?" message

### ISSUE THREAD

**Issue by coffeemug** [rethinkdb/rethinkdb#4754]
The error message is really confusing because it doesn't tell the user anything about what's going on (what are "start stamps?")

We should rephrase it as something like this: "Changefeed has been interrupted by resharding."

**Comment by coffeemug** [comment]
Also, /cc @danielmewes @Tryneus since we talked about this internally.

**Comment by mlucy** [comment]
I think there are other failure modes for this besides resharding.  I agree that talking about start stamps is probably bad.  How about something like "Changefeed setup failed.  Are you in the middle of resharding?".

**Comment by mglukhovsky** [comment]
@mlucy: what are the other failure modes?

**Comment by mlucy** [comment]
Anything that can cause a shard to lose its changefeed server between sending the `changefeed_subscribe_t` and sending the `changefeed_stamp_t`.  It's possible that resharding is the only thing that can cause that, but I'm not sure.  It should always be caused by a cluster event, though.  (For example, I wouldn't be shocked if there was a specific timing scenario where a server hosting a shard crashing and then quickly coming back up could cause this, although I think that will probably always trigger a "changefeed disconnected" error instead.)

Theoretically you can also get that error if the changefeed client that's trying to do the stamp somehow unregistered itself from the changefeed server before the stamp arrived, but I think if that happens the user should get an error from whatever unregistered it instead of this error.

**Comment by coffeemug** [comment]
> Changefeed setup failed. Are you in the middle of resharding?

Does this happen to new changefeeds or existing ones? If it happens to new ones, this makes sense. Otherwise, I'd prefer the following instead:

> Changefeed has been interrupted. Are you in the middle of resharding?

**Comment by mlucy** [comment]
This particular error can only occur when a changefeed is first being set up.

**Comment by danielmewes** [comment]
I think "Changefeed setup failed. Are you in the middle of resharding?" is a good error message then.

**Comment by coffeemug** [comment]
:+1:

**Comment by williamstein** [comment]
(Sorry for sticking basically a support question into the middle of an issue -- but maybe this is relevant since it illustrates a case where I -- as a user -- see the error message this thread is about.)

My current SageMathCloud production setup involves 6 client node.js servers and 6 rethinkdb nodes.  (See attached screenshots below.)  As a "drill", with very low load (it's saturday morning), I just tried typing "sudo reboot" on db5 (one of the rethinkdb nodes).   The result was that in about 1-2 minutes I had thousands of  `{"name":"ReqlOpFailedError","msg":"Unable to retrieve the start stamps.  Did you just reshard?",...` in the log files for all of my clients.   Waiting a few more minutes didn't help at all.     Due to every changefeed failing, my clients (I guess stupidly) created too many new connections before the old ones terminated -- that's my fault, and I should improve my connection pool.     

To fix the problem I manually restarted all the rethinkdb nodes and clients.   This worked. 

In any case, right now if even a single one of my rethinkdb nodes goes down for any reason (even intentional), it's a total disaster due to "Unable to retrieve the start stamps".   Running things this way is very bad.      If this is not a bug in RethinkDB, do you have any suggestions about how I should change my application?   E.g., do you think it would have if whenever the application sees   ``{"name":"ReqlOpFailedError","msg":"Unable to retrieve the start stamps. ` it maybe just waits k seconds (?) before attempting to create any more changefeeds?       

Of course this whole problem arises because my application code has to deal with automatically recreating changefeeds when the feed fails, since that's not builtin to RethinkDB yet.   If that's what you guys are actively working on right now, maybe my best strategy would be to just wait until you're done with that.

<img width="560" alt="screen shot 2015-09-26 at 6 43 03 am" src="https://cloud.githubusercontent.com/assets/1276278/10117829/de1331c8-6419-11e5-8131-831fa6a5aab8.png">
<img width="890" alt="screen shot 2015-09-26 at 6 42 53 am" src="https://cloud.githubusercontent.com/assets/1276278/10117830/de16ff42-6419-11e5-85b8-4f39250ec49e.png">

**Comment by coffeemug** [comment]
Hi @williamstein, thanks for reporting this!

Theoretically, this is expected behavior -- it's very difficult to keep changefeeds running efficiently across machine failures, so we designed them in such a way that they need to be restarted in case of hardware failure.

In practice, I can definitely see how this makes building resilient applications very difficult (especially the part where your code can't just reconnect, but has to wait for failover to happen). Issue #3471 should partially address this problem (it doesn't deal with server failures, but network interrupts between client and server). As we work on it, we'll look into how to make this case more robust, or at least easier to deal with.

In the meantime, I'd suggest this -- your code can call `wait` on a table to ensure the table is ready (see http://rethinkdb.com/api/javascript/wait/). Before you set up your feeds in the client code, consider calling `wait` -- this way you won't get a cascading failure when you try to restart the clients immediately after a server failure.

We'll try hard to address this during the work on #3471 in some way; we'll either make it easier, or give better guidance/sample code in the docs, or both.

**Comment by williamstein** [comment]
Awesome, thanks so much for the suggestion!  I'll modify my code to use wait, and also report back with what happens in practice.  Thanks!

**Comment by williamstein** [comment]
I implemented your wait suggestion right now, and hit what might be a mistake in the documentation (I'm using the Javascript driver from node.js).  The documentation says `waitFor: a string indicating a table status to wait on before returning, one of ready_for_outdated_reads, ready_for_reads, ready_for_writes, or all_replicas_ready. The default is ready_for_writes.`    However, when I call wait without specifying a `waitFor` parameter, it seems to me like it waits for `all_replicas_ready`, i.e., it seems like the documentation is wrong (or there is a bug).      Aside: it's awesome how ridiculously easy it is to make a "cluster" of 3 rethinkdb nodes on the same computer just by editing some files in /etc/rethinkdb/ !]  

It seems logical to me to just use 'ready_for_reads' since I only need to read from the tables involved in the changefeeds.  (Even 'ready_for_outdated_reads' would be fine for my needs.)  However, when I tried 'ready_for_reads'  with my single-machine 3-node test "cluster", killing/starting individual rethinkdb servers, I did suddenly get a lot of `Unable to retrieve the start stamps.  Did you just reshard?` messages, despite the admin interface saying everything was fine.  

I then switched to using 'ready_for_writes' explicitly (with `waitFor`), and tried killing/starting/stopping in many different orders and ways, with around 100 changefeeds.  With my testing cluster everything then WORKED PERFECTLY.

**Comment by williamstein** [comment]
Additional followup.  I tried using my new code (with waitFor:'ready_for_writes') in production, testing restarting when node.    It didn't really work in that the wait started passing fine, but I still started getting "Did you just reshard?" errors.   The server would stupidly just keep retrying (dozens of times a second on various changefeeds), and I think that made the server never finish whatever it needed to do so as to not return the "Did you just reshard?" error.  So I changed my code so that (in addition to the wait code), also if when making a changefeed there's an error message that contains the string "Did you just reshard?", then my code waits a few seconds (currently 10s) before returning the error.  This reduces the load on the server and this did actually work.   There are still some "Did you just reshard?" errors, and it takes a while for everything to stabilize, but it 100% does work without any intervention.  This is a massive improvement over the situation before.  

It would of course be way better if wait(waitFor='ready_for_writes') would only return -- "sure the table is ready to go!" -- if I'm not going to the subsequently get "Did you just reshard?" errors.   If that were the case, I could delete that hack I put in involving the delay in case of a ""Did you just reshard?" message, and also things would probably just work much better.     I guess what I really want is  wait(waitFor='ready_for_changefeeds')...

**Comment by coffeemug** [comment]
@williamstein -- I believe `all_replicas_ready` is a stronger guarantee than `ready_for_writes`, and that's the one you might need not to get the "did you reshard?" error after the wait. Could you try this mode without the hack and see if that works?

I'll talk to some folks on Mon to clarify the difference between different states in the docs, and make sure we document the right default.

**Comment by coffeemug** [comment]
Also, if `all_replicas_ready` doesn't guarantee successful feed creation, that is a bug and we will fix it.

**Comment by williamstein** [comment]
On Sat, Sep 26, 2015 at 10:32 PM, Slava Akhmechet <notifications@github.com>
wrote:

> Also, if all_replicas_ready doesn't guarantee successful feed creation,
> that is a bug and we will fix it.
>
>
Thanks!

What is the definition of "all_replicas_ready"?  In particular, suppose I
have a cluster with 3 nodes and every table has 3 replicas.  Then I kill
one of the nodes and wait five minutes (say).  Is "all_replicas_ready"
supposed to be true then (since node 3 will have been removed completely)?
       In particular, should automatic failover in the case of losing one
node (assuming I have enough replicas) still work and allow my changefeeds
to work again?

William


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rethinkdb/rethinkdb/issues/4754#issuecomment-143521087>
> .
>



-- 
William (http://wstein.org)

**Comment by coffeemug** [comment]
Err, I think my previous message was wrong. The definition of `all_replicas_ready` is that *all* the replicas for a given table are ready, which means it wouldn't work in case of server failure.

Let me chat with some folks on Monday who *actually* know all the edge cases of this stuff, and report back. We'll also improve the docs to make everything more clear.

**Comment by VeXocide** [comment]
@williamstein -- I'll give this a shot, but note that I might be insufficiently caffeinated.

`ready_for_writes` is relatively straightforward, it takes the configuration the system is running on and verifies that for each configured shard the primary is available, and a majority of the replicas. This essentially reports whether, had a write been issued at the time, it would have successfully made it to disk and be acknowledged.

Unfortunately this is insufficient for your changefeeds case because it is based on the configuration the system is currently using, not the configuration the system is configured to use. In other words, it will not wait for a configuration changes to be applied, which include rebalancing, and reconfigurations that change the sharding and replication factors.

If you were to watch the `ready_for_writes` during a configuration change it'd start off as true. It would then change to false for a moment as the primary and secondaries restart with the new configuration, be false while a majority is unavailable due to backfilling if this needed to happen due to shard boundaries changing. Subsequently it would become true again.

Next is `all_replicas_ready` whose name unfortunately doesn't do it's justice, and there's an issue discussing the renaming of it, https://github.com/rethinkdb/rethinkdb/issues/4334. This mode verifies that every replica is available for each shard, and that all the configurations match the latest configuration set by the user. I'm still of the opinion that `running_as_configured` would be a better name.

This would work for your changefeed case because, contrary to `ready_for_writes` above, it does wait for configuration changes to be applied and the system to be running using the latest configuration set by the user. However, it is likely too strong as it requires every replica to be available for each shard. As such would report false when the primary and a majority, but not every replica, is available and running using the latest configuration.

If you were to watch the `all_replicas_ready` during a configuration change it would become false the moment the change is accepted by raft. It would then remain false until all shards and replicas were running with that configuration.

Now it's once again become clear why I'm a systems engineer and not a technical writer -- it's surprisingly hard to describe things in a way that's both clear and technically correct. At least I hope this is clear, please let me know if it's not.

Finally it's time to actually answer your questions. Using your case where you have a cluster of three nodes, and every table consists of one shard and three replicas. When you kill one of the nodes `all_replicas_ready` will remain false until one of two things happen. It will, obviously, become true again when it's restarted, and it will also become true again when the configuration is changed to no longer include the replica that was killed.

The cluster would however function fine, that is it would select a new primary through the automatic failover and be available for writes and everything. I'm no expert on changefeeds (@mlucy is, and please correct me if I'm wrong) but they should work fine as well.

What you really need is a mode that verifies whether the primary and a majority of replicas is available, and running the latest configuration set by the user. Unfortunately we don't have such a mode at the moment, but we should be able to add it.

Hope this helped, now it's time for some coffee.

**Comment by williamstein** [comment]
> Also, if all_replicas_ready doesn't guarantee successful feed creation, that is a bug and we will fix it.

Then there might be a bug:  Below I call wait with all_replicas_ready on my 3-node test cluster and it says `{ ready: 1 }`.  Then I try to create a changefeed, and get `Unable to retrieve the start stamps.  Did you just reshard?`.   I waited 5 minutes, with some other clients running, probably also periodically trying to create changefeeds and with connections open, and the situation doesn't improve.  I then restarted my other clients and everything works fine.  So somehow existing open connections seem to be problematic/relevant (?). 

```
coffee> db.table('projects').wait(waitFor:'all_replicas_ready', timeout:3).run(console.log)
2015-09-28T00:48:17.585Z - debug: rethink: query -- (1 concurrent) -- 'r.db("smc").table("projects", {readMode: "outdated"}).wait({waitFor: "all_replicas_ready", timeout: 3})'
{ _bitField: 1,
  _fulfillmentHandler0:
   { [Function: runBound]
     domain:
      { domain: null,
        _events: [Object],
        _maxListeners: undefined,
        members: [] } },
  _rejectionHandler0:
   { [Function: runBound]
     domain:
      { domain: null,
        _events: [Object],
        _maxListeners: undefined,
        members: [] } },
  _progressHandler0: undefined,
  _promise0: [Function],
  _receiver0: [Circular],
  _settledValue: undefined }
coffee> 2015-09-28T00:48:17.603Z - debug: rethink: query time using (ddd4d90a-fe87-406f-b656-2c8c12fd2046) took 18ms; average=16ms; 0 concurrent -- 'r.db("smc").table("projects", {readMode: "outdated"}).wait({waitFor: "all_replicas_ready", timeout: 3})'
undefined { ready: 1 }
 
undefined
coffee> db.table('projects').changes().run(console.log)
2015-09-28T00:48:19.502Z - debug: rethink: query -- (1 concurrent) -- 'r.db("smc").table("projects", {readMode: "outdated"}).changes()'
{ _bitField: 1,
  _fulfillmentHandler0:
   { [Function: runBound]
     domain:
      { domain: null,
        _events: [Object],
        _maxListeners: undefined,
        members: [] } },
  _rejectionHandler0:
   { [Function: runBound]
     domain:
      { domain: null,
        _events: [Object],
        _maxListeners: undefined,
        members: [] } },
  _progressHandler0: undefined,
  _promise0: [Function],
  _receiver0: [Circular],
  _settledValue: undefined }
coffee> 2015-09-28T00:48:19.507Z - debug: rethink: query time using (ddd4d90a-fe87-406f-b656-2c8c12fd2046) took 6ms; average=15ms; 0 concurrent -- 'r.db("smc").table("projects", {readMode: "outdated"}).changes()'
{ [ReqlOpFailedError: Unable to retrieve the start stamps.  Did you just reshard? in:
r.db("smc").table("projects", {readMode: "outdated"}).changes()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^]
  name: 'ReqlOpFailedError',
  msg: 'Unable to retrieve the start stamps.  Did you just reshard?',
  frames: [],
  message: 'Unable to retrieve the start stamps.  Did you just reshard? in:\nr.db("smc").table("projects", {readMode: "outdated"}).changes()\n^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^' } undefined
```

**Comment by danielmewes** [comment]
`"ready_for_reads"` is enough to guarantee successful changefeed creation. The reason it currently isn't is https://github.com/rethinkdb/rethinkdb/issues/4838 which we're fixing in the next point release.

**Comment by williamstein** [comment]
Awesome -- I am eagerly awaiting the next point release!

**Comment by danielmewes** [comment]
Yeah, really sorry for your bumpy experience William! Everything should be much smoother soon.�

**Comment by williamstein** [comment]
Thanks.  (Even with bumps, writing code against RethinkDB is awesome;  I'm so glad you guys created this!  It's definitely what I've always wanted in a database.)

**Comment by coffeemug** [comment]
@danielmewes @mlucy -- I think the `changes` command should internally wait for `ready_for_reads` when it's first called.

We don't typically wait on other queries, because if the cluster isn't ready and the query fails, the end user can just "rerun" it by refreshing the page, clicking a button again, etc. This way the developer can pretty much ignore the issue in the code (other than handling the exception and propagating the error message to the end user).

However, this isn't the case with changefeeds. If a node fails, the developer has to recreate the feed, and therefore any time anybody uses the `changes()` command, the proper way to write the code is to prepend it with a `wait`. This sounds unnecessarily painful and complicated -- the user shouldn't have to do that. Which is why I think `changes()` should block at the beginning to make sure the cluster is ready. Opinions?

**Comment by coffeemug** [comment]
Note for all you e-mail users, I edited the comment above to make it more clear; please view in GitHub.

**Comment by mlucy** [comment]
@coffeemug -- this seems way outside the original scope of improving an error message, could we open a new issue for this?

**Comment by coffeemug** [comment]
Yes -- see #4893.
